### DISCUSSION TITLE
Add new style property / update text-offset to better support variable anchor placement?

### DISCUSSION THREAD

**Discussion by drwestco** [maplibre/maplibre-gl-js#1832] [Ideas]
text-offset and text-radial-offset both seem tailored to placement of labels around a center-anchored icon. For pushpins, teardrops, or other asymmetrical icons, I don't see style syntax to distribute labels evenly around the icon's perimeter.

e.g., if the desired label placements for all eight variable-anchor positions are as follows, what's the best way to indicate this through the symbol layer style? The stop sign icon is anchored at the bottom, so the desired text-offset for 'left' vs 'bottom-left' would have a different y value.

<img width="326" alt="image" src="https://user-images.githubusercontent.com/743030/200913484-e642cdbf-17c3-4df3-8ae9-50b0906f3e96.png">

Assuming I'm correct the current style schema doesn't support this, what would a new or modified style property look like? We could be completely explicit and have something like an offset per text anchor, but maybe there's a more elegant solution?

```
'text-anchor-offset': {
  'top': [x1, y1],       // [x1, y1] is used when the 'top' value is picked from 'text-variable-anchor'
  'top-left': [x2, y2],  // [x2, y2] is used when the 'top-left' value is picked from 'text-variable-anchor'
 ...
}

**Comment by HarelM** [comment]
I'm not sure I'm following. Can you give an example of something that you currently can't do? Something very specific.
The above description is a bit too vague for me...

**Reply by drwestco** [reply]
Does the image help? If the desire is to have all eight of the green positions available for the control to choose from, or even just four:

```
"text-variable-anchor": ["top", "right", "bottom", "left"]
```

A single x/y pair for `text-offset` isn't sufficient as the desired Y value for the `right` and `left` positions is different from that for `top` and `bottom`.

`text-radial-offset` also doesn't help, since the distance from the anchor point (yellow dot) is different for each of the possible positions.

**Reply by HarelM** [reply]
But then you need multiple values for the text as well, and maybe font, and size. Wouldn't this explode?

**Reply by drwestco** [reply]
I don't follow. Why would the font/size need to change based on where the label for an icon is placed?

**Reply by drwestco** [reply]
That said, secondary labels with different font style and independent hover/click behavior would be an excellent feature, but that's separate from this request.

![image](https://user-images.githubusercontent.com/743030/201737985-27c1466d-9082-4fe7-b3b1-c09dab5fce00.png)

**Reply by HarelM** [reply]
If I understand the question/feature/requirement your are saying that a single place for text isn't enough.
Let's, for the sake of the discussion, say we have two places for the text: below and above the icon.
In that case you'd want to present different text, right? Otherwise, what's the point of placing the same text in two different places?

**Reply by drwestco** [reply]
No. `text-variable-anchor` (existing style) already provides **alternative** positions for the text, not **additional** ones. If the text fits below the icon, it shows up below. If something else is occupying that space, the control tries the other positions specified by `text-variable-anchor`.

**Reply by HarelM** [reply]
Is this the same as here https://github.com/mapbox/mapbox-gl-js/issues/8598?

**Reply by drwestco** [reply]
That issue touches on it, but doesn't completely address it. The discussion + solution (`text-radial-offset` and two-element `text-offset`) still assumes a symmetric icon, centered over its anchor. It fails to address teardrop-style icons with the icon anchor at the tip.

**Reply by HarelM** [reply]
I can't say I fully understand all the bit and bytes here unfortunately.
I'm not sure what was merged in the PR linked in the issue above as it talks about a rectangle you can provide instead of radial offset (?), along with adding a text-offset to move the rectangle upwards so that it wraps the icon in a symetrical way, this should solve your issue, wouldn't it?
I'm just not sure what was merged and how to use it.

**Reply by drwestco** [reply]
From what I can tell with the discussion, they're describing how `text-variable-anchor`, `text-radial-offset` and `text-offset` work together. This is the same in MapBox and MapLibre. You don't get to specify a rectangle - there's an implicit rectangle defined by the x/y values from `text-offset`. The rectangle is still centered on the point geometry. There's currently no syntax (that I know of) to shift that rectangle to account for an asymmetric icon.

**Reply by HarelM** [reply]
Can you draw the rectangle made by the 3 properties? Or link to an example with debug frame stuff turned on? I'm sorry I keep pulling you back, I just want to better understand all this...

**Comment by drwestco** [comment]
Example using `"text-radial-offset": 1`. The nearest edge of the text is placed 1em from the anchor point (center of icon).

![image](https://user-images.githubusercontent.com/743030/201750636-6961c05f-e0db-4f1c-98b0-f404f97afe27.png)

Example using `"text-offset": [4, 1]`. When the left or right anchor position is chosen, the nearest edge of the text is placed 4em from the anchor point. When top or bottom is chosen, the nearest edge is placed 1em from the anchor point.

![image](https://user-images.githubusercontent.com/743030/201751119-7311f14f-46fc-43b8-8426-5abf6fd84e67.png)

It seems like you can then use `icon-offset` to shift the icon and `text-translate` to shift the text to match. But as the latter is a paint property, collisions aren't properly considered. Text for 'District Drum Tribe' overlaps icon for 'Motown Memories'.

![image](https://user-images.githubusercontent.com/743030/201754191-c37c7c8e-2098-43b2-a331-71ccfa3a4bf4.png)

**Reply by HarelM** [reply]
I had a feeling there's a way to move the text etc. It might be better to focus on fixing the collision for text offset and icon offset instead of trying to invent another property, wouldn't it?

**Reply by drwestco** [reply]
Yes, updating existing styles would be preferable, if possible. Thus the discussion instead of a PR. :-)

I don't think `text-translate` helps, by design. Paint properties explicitly don't factor into collisions.

**Reply by HarelM** [reply]
If anything, text-offset is the most unintuitive for me, the two numbers representing different offset according to position is "not readable".
But everything in the style spec is arrays, so if we introduce a new "type of array" (or array with more numbers) to text-offset we can facilitate for this request, can't we?

**Reply by drwestco** [reply]
Yeah, extending text-offset should do. It's just a matter of what the new syntax looks like. It could be an array of x/y pairs, one for each entry in text-variable-anchor; it could be a structured map like my original example; something else?

**Reply by HarelM** [reply]
I don't think the style spec supports structured maps, only "weird" arrays.
Since I don't like the mixed string and number arrays I would vote for array of two numbers array, and depending on the size (like in css border etc) act accordingly.

**Reply by drwestco** [reply]
Sounds reasonable. I'll write up my proposed changes for text-offset as a new thread in this discussion.

**Comment by drwestco** [comment]
### Point
A point is a two-element array of numbers, representing an [x, y] coordinate pair. Units and other interpretation of the x/y values depend on where and how the point type is used.

### text-offset
_Layout property. Optional point or array of points. Units in ems. Defaults to [0,0]. Requires text-field. Disabled by text-radial-offset. Supports interpolate expressions._

Offset distance of text from its anchor. Positive values indicate right and down, while negative values indicate left and up. 

If used with _text-variable-anchor_, a single point used for _text-offset_ will be taken as absolute values. Offsets along the x- and y-axis will be applied automatically based on the anchor position.

If an array of points is provided for _text-offset_, the number of points should match the number of entries provided via _text-variable-anchor_. Each point in the array is used as the offset distance of text from its anchor, when the corresponding anchor is used by the renderer. If more entries are provided for _text-variable-anchor_ than _text-offset_, the last value in _text-offset_ will be used.

**Example**
```
{
    "text-offset": [[0, 4], [3, 0]],
    "text-variable-anchor": ["top", "left"]
}
```
When the renderer chooses the _top_ anchor, [0, 4] will be used for _text-offset_; the text will be shifted down by 4 ems. When the renderer chooses the _left_ anchor, [3, 0] will be used for _text-offset_; the text will be shifted right by 3 ems.

**Reply by HarelM** [reply]
This looks great! Thanks for putting this into words.
A thought that crossed my mind after reading this - if the renderer chooses a location, but then sees that the offset causes symbol collision, what is the expected behavior?

**Reply by drwestco** [reply]
It's described in the docs for text-variable-anchor:

To increase the chance of placing high-priority labels on the map, you can provide an array of text-anchor locations: the renderer will attempt to place the label at each location, in order, before moving onto the next label.

"location" here means anchor + text-offset.

**Reply by drwestco** [reply]
This also ties into the Style Validation conversation. We really need some way to represent properties with alternative type signatures.
